server: fix scheduling group update timing in system.clients

calebxyz opened the pull request:
Previously, the scheduling_group column was updated during the switch_tenant function, which meant the update occurred only after the tenant change operation completed—updating rows one by one. With this change, the scheduling_group column is now updated before the switch_tenant logic runs, ensuring that the table reflects the correct scheduling groups for all rows as early as possible.

fixes: #26060
fixes: #27295

backport: not required 
this is a minor bug fix. Internal logic worked but the user couldnt see the change in the table if they would read the system.clients table

piodul commented on transport/server.cc:
```diff
@@ -992,11 +992,18 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
-    switch_tenant([this] (noncopyable_function<future<> ()> process_loop) -> future<> {
-        auto shg = co_await _server._sl_controller.get_user_scheduling_group(_client_state.user());
-        _current_scheduling_group = shg;
-        co_return co_await _server._sl_controller.with_user_service_level(_client_state.user(), std::move(process_loop));
+    auto update_ret = update_scheduling_group_impl();  
```
You can't preempt the currently executing task in a destructor (with the exception of `seastar::async`, but just not in general), so this would require blocking the whole thread/shard - most likely indefinitely, because our future/promise pairs are not thread safe and cannot migrate between threads, so no other thread would be able to unblock the thread that waits on the future.

piodul commented on test/cqlpy/test_service_level_api.py:
```diff
@@ -12,25 +12,35 @@
 from .util import new_session, unique_name
 import time
 
+
```
Same goes for the whitespace changes in this file.

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -634,17 +634,31 @@ future<scheduling_group> service_level_controller::auth_integration::get_user_sc
         co_return _sl_controller.get_scheduling_group(sl_name);
-    }
-    else {
+    } else {
```
This is a stylistic cleanup unrelated to the fix. Consider doing it in a separate commit.

piodul commented on transport/server.cc:
```diff
@@ -994,21 +995,32 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
+}
+
+void cql_server::connection::update_user_scheduling_group_v1(std::optional<auth::authenticated_user>& usr) {
+    switch_tenant([this, usr] (this auto&& self, noncopyable_function<future<> ()> process_loop) -> future<> {
```
I can still see `auto&&` here

calebxyz replied:
we cant use && here because only when we pass self as value it will copy all the capture to the lambda and fix the fiasco problem

piodul commented on transport/server.cc:
```diff
@@ -995,20 +997,38 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
 }
 
+void cql_server::connection::update_user_scheduling_group_v1(const std::optional<auth::authenticated_user>& usr) {
+    switch_tenant([this, usr] (this auto&& self, noncopyable_function<future<> ()> process_loop) -> future<> {
```
`this auto&&`, I think this was supposed to be without `&&`. I'm not sure what was the point of using deducing this here (I'm not very familiar with this feature), but I know that we started using it to avoid the lambda coroutine fiasco (https://docs.seastar.io/master/tutorial.html#lambda-coroutines-c-23-and-later), and this requires passing `this` by value.

calebxyz replied:
correct this is a mistake, and capuring this is wrong now

piodul commented on transport/server.cc:
```diff
@@ -12,7 +12,9 @@
 #include <seastar/core/semaphore.hh>
+#include "seastar/core/sstring.hh"
 #include "seastar/coroutine/switch_to.hh"
+#include "transport/event.hh"
```
Nit: not sure if adding these hearders is strictly necessary. Moreover, we include seastar headers via angle brackets.

piodul commented on test/cqlpy/test_service_level_api.py:
```diff
@@ -107,47 +153,51 @@ def test_count_opened_cql_connections(cql):
+# while using the cache from group0 this test will only work with raft clusters 
+# firt we will create a new service level and attach it then query it to see 
+# taht all connections are updated immidiatly 
+def test_appending_new_service_level_no_raft(cql):
```
There are two functions called `test_appending_new_service_level_no_raft` in this file. I suppose that this one should be called e.g. `test_appending_new_service_level_with_raft`?

piodul commented on test/cqlpy/test_service_level_api.py:
```diff
@@ -107,47 +153,51 @@ def test_count_opened_cql_connections(cql):
+    try:
+        with new_session(cql, user) as user_session:
+            wait_for_clients_and_attach_service_level(cql, user, sl1, sl2, shard_count)
+            verify_scheduling_group_assignment(cql, user, sl2, shard_count)
```
To me, it looks like the only difference between this test and `test_appending_new_service_level_no_raft` (the other one) is that we don't do read barriers here. Perhaps we could reuse the previous test surround `for _ in range(100): read_barrier(user_session)` with an `if`?

piodul replied:
By the way, I am not sure that we are actually running cqlpy tests both in gossip topology mode and raft topology mode (this affects whether v1 or v2 is used). The behavior tested in cqlpy usually doesn't depend on this, so we might be running this in one mode only (I suspect it might be just raft topology).

If you want to test it in two modes, you will have to move this test to a different framework which allows you to customize which mode will be used. We actually have some service level tests in `test/cluster/auth_cluster/test_raft_service_levels.py` (though if you want to test non-raft topology then perhaps a new file should be used, or the file should be renamed). Look for the use of `force_gossip_topology_changes` in `test/cluster/` directory to see how to start a cluster which uses gossip topology .

calebxyz commented on test/cqlpy/test_service_level_api.py:
```diff
@@ -107,47 +153,51 @@ def test_count_opened_cql_connections(cql):
-        cql.execute(f"DROP ROLE {user}")
-        cql.execute(f"DROP SERVICE LEVEL {sl1}")
-        cql.execute(f"DROP SERVICE LEVEL {sl2}")
+        finalize_service_level_test(cql, user, [sl1, sl2])
```
the reformat of this test was a direct effect of the changes that were introduced to the code, do we still need to create new commit for that?

calebxyz commented on transport/server.cc:
```diff
@@ -995,20 +997,38 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
-            _current_scheduling_group = shg;
-            co_return co_await _server._sl_controller.with_user_service_level(_client_state.user(), std::move(process_loop));
-        });
+        update_user_scheduling_group_v1(_client_state.user());
```
i think its better to do it at once in this case but noted

calebxyz commented on test/cqlpy/test_service_level_api.py:
```diff
@@ -110,44 +110,3 @@ def test_count_opened_cql_connections(cql):
-
-# Test if `/service_levels/switch_tenants` updates scheduling group 
-# of CQL connections without restarting them.
-# 
```
its duplication , i have moved it to clusters to test gossip and raft modes. in this area we can only test raft from what i understand in cluster test we can test both modes and for now we have them both so i thought its a better test.

avikivity commented:
@andrzej-jackowski-scylladb please review

andrzej-jackowski-scylladb commented on transport/server.cc:
```diff
@@ -995,20 +995,38 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
-            _current_scheduling_group = shg;
-            co_return co_await _server._sl_controller.with_service_level(qos::service_level_controller::driver_service_level_name, std::move(process_loop));
-        });
+        update_service_level_scheduling_group();
```
The `update_service_level_scheduling_group` is misleading. If it's only for control connection better name it `update_control_connection_scheduling_group`

andrzej-jackowski-scylladb reviewed (changes requested):
The first commit (`server: Refactor scheduling group update timing in system.clients`) mixes refactoring (moving the `update_scheduling_group` logic into new functions) with the fix for #26060. These should be split into two separate commits.

calebxyz commented:
> The first commit (`server: Refactor scheduling group update timing in system.clients`) mixes refactoring (moving the `update_scheduling_group` logic into new functions) with the fix for #26060. These should be split into two separate commits.

We had a discussion with @piodul about it and decided it better to be in the way it is now

piodul commented:
> > Well, if the changes are small enough then the split is, perhaps, not necessary, but I agree here that due to the volume of other changes the details about `shg_grp` and `this auto self` can easily be missed. I guess that is subjective but let's err on the side of smaller commits.
> 
> i dont see a reason for a split, but as i mention the self deducing here is not important and i can remove, if you think its better i can create a small commit for it. please pick

Ok - in that case please split the first commit.

calebxyz commented:
V3: 1.Split down commits to show progress as was suggested.
2.removed self deducing this in lambda where it's not needed.

@piodul @andrzej-jackowski-scylladb please review

andrzej-jackowski-scylladb commented on transport/server.cc:
```diff
@@ -1009,19 +1009,27 @@ void cql_server::connection::update_scheduling_group_v1() {
+        auto shg_grp = _server._sl_controller.get_cached_user_scheduling_group(_client_state.user());
+        _current_scheduling_group = shg_grp;
+        switch_tenant([this] (this auto self, noncopyable_function<future<> ()> process_loop) -> future<> {
+        co_return co_await _server._sl_controller.with_user_service_level(_client_state.user(), std::move(process_loop));
```
Is the indentation here broken intentionally?

andrzej-jackowski-scylladb commented on transport/server.cc:
```diff
@@ -993,6 +993,22 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
     co_return res;
 }
 
+void cql_server::connection::update_scheduling_group_v1() {
```
73a724b4aa8388f3edf9f542669e7ffa5891d9fd commit message:
 - the commit message mentions `update_service_level_scheduling_group` and  `update_user_scheduling_group_v1`, whereas in fact only `update_scheduling_group_v1` is added
 - placehholders -> placeholders

andrzej-jackowski-scylladb commented on transport/server.cc:
```diff
@@ -995,9 +995,9 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
 
 void cql_server::connection::update_scheduling_group_v1() {
     if (_client_state.is_control_connection()) {
+        auto shg_grp = _server._sl_controller.get_scheduling_group(qos::service_level_controller::driver_service_level_name);
```
Why we are changing the logic in `update_scheduling_group_v1`?
Based on the commit message, I thought only `v2` will be fixed. That's also how I understood @piodul's design: `otherwise, fallback to the current implementation which looks up the scheduling group inside the lambda passed to switch_tenant`.

calebxyz replied:
thinking about it maybe we should change it in V1 as well to make it perfectly aligned

andrzej-jackowski-scylladb commented on transport/server.cc:
```diff
@@ -993,43 +993,37 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
-        });
-    }
+void cql_server::connection::update_user_scheduling_group_v1(const std::optional<auth::authenticated_user>& usr) {
+    switch_tenant([this, usr] (this auto self, noncopyable_function<future<> ()> process_loop) -> future<> {
```
Again, `this auto self` is added in the refactoring commit.
Is it really needed? I think currently there are 0 uses of `this auto self` in the entire ScyllaDB repo.
If it's needed please explain in the commit message.

andrzej-jackowski-scylladb commented on transport/server.cc:
```diff
@@ -993,22 +993,41 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
     }
 }
 
+
```
nit: unneeded space

andrzej-jackowski-scylladb commented on transport/server.cc:
```diff
@@ -993,22 +993,41 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
-        });
+        update_control_connection_scheduling_group();
+    } else if (_server._sl_controller.is_v2()) {
+        update_user_scheduling_group_v2(_client_state.user());
```
I have one serious concern here. Does `_sl_controller.is_v2() == true` always imply `_sl_controller._sl_data_accessor->can_use_effective_service_level_cache() == true`?

I can see the following code:
```
future<std::optional<service_level_options>> service_level_controller::auth_integration::find_effective_service_level(const sstring& role_name) {
    const auto _ = _stop_gate.hold();

    if (_sl_controller._sl_data_accessor->can_use_effective_service_level_cache()) {
        auto effective_sl_it = _cache.find(role_name);
        co_return effective_sl_it != _cache.end() 
            ? std::optional<service_level_options>(effective_sl_it->second)
            : std::nullopt;
    } else {
```

Therefore my understanding is: if `_sl_controller._sl_data_accessor->can_use_effective_service_level_cache() == true` we can just use the cache, that's what `find_effective_service_level` already did.

However, `can_use_effective_service_level_cache` is implemented as follows:
```
bool raft_service_level_distributed_data_accessor::can_use_effective_service_level_cache() const {
    return !auth::legacy_mode(_qp);
}
```

@calebxyz, @piodul, @Jadw1, is it really safe to rely on cache (in `cql_server::connection::update_scheduling_group`) if `_sl_controller.is_v2() == true`  or there is some case (e.g. during auth or service level upgrade) when it makes a difference and `can_use_effective_service_level_cache` should be checked?

calebxyz replied:
> This will most likely require introducing `can_use_effective_service_level_cache` into the service level controller, e.g.:

Will add it on Sunday, good catch @andrzej-jackowski-scylladb

Jadw1 commented on transport/server.cc:
```diff
@@ -1010,9 +1010,9 @@ void cql_server::connection::update_user_scheduling_group_v2(const std::optional
 }
 
 void cql_server::connection::update_control_connection_scheduling_group() {
+    auto shg_grp = _server._sl_controller.get_scheduling_group(qos::service_level_controller::driver_service_level_name);
```
What if we are not using consistent topology and service levels on Raft?

piodul commented:
Queued
